# Price Cut Chandler Random Forest

### Data Cleanup

The following property types were eliminated from the original CSV:
1. Lot/Land

In [1]:
# Dependencies
import pandas as pd
from pathlib import Path

In [2]:
# Name of the CSV file
file_path = Path("Resources/Zillow_Chandler_For_Sale.csv")
df = pd.read_csv(file_path)
df

Property URL  MLS ID  \
0    https://www.zillow.com/homedetails/93-N-Cooper...     NaN   
1    https://www.zillow.com/homedetails/295-N-Rural...     NaN   
2    https://www.zillow.com/homedetails/1351-N-Plea...     NaN   
3    https://www.zillow.com/homedetails/1075-E-Chan...     NaN   
4    https://www.zillow.com/homedetails/295-N-Rural...     NaN   
..                                                 ...     ...   
434  https://www.zillow.com/community/village-at-ha...     NaN   
435  https://www.zillow.com/homedetails/960-E-Mulbe...     NaN   
436  https://www.zillow.com/community/willis-common...     NaN   
437  https://www.zillow.com/homedetails/972-E-Mulbe...     NaN   
438  https://www.zillow.com/community/village-at-ha...     NaN   

    Listing description Property type  Number of days on Zillow  \
0    Apartment for sale     Apartment                       NaN   
1    Apartment for sale     Apartment                       5.0   
2    Apartment for sale     Apartment                       4.0   
3    Apartment for sale     Apartment                      15.0   
4    Apartment for sale     Apartment                       NaN   
..                  ...           ...                       ...   
434    New construction       Unknown                      41.0   
435    New construction       Unknown                      74.0   
436    New construction       Unknown                       NaN   
437    New construction       Unknown                      74.0   
438    New construction       Unknown                      41.0   

     Property price (USD) Price was cut  Price cut amount (USD)  \
0                  289000            No                     NaN   
1                  389999            No                     NaN   
2                  398000            No                     NaN   
3                  225000            No                     NaN   
4                  299900           Yes                 35100.0   
..                    ...           ...                     ...   
434                607000            No                     NaN   
435                607000            No                     NaN   
436                580000            No                     NaN   
437                595000            No                     NaN   
438                595000            No                     NaN   

    Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                           NaN         1075  ...        2          2   
1                           NaN         1069  ...        2          2   
2                           NaN         1480  ...        3          2   
3                           NaN          705  ...        1          1   
4                    02/13/2024          979  ...        1          1   
..                          ...          ...  ...      ...        ...   
434                         NaN         1324  ...        3          2   
435                         NaN         1324  ...        3          2   
436                         NaN         2101  ...        3          3   
437                         NaN         1292  ...        3          2   
438                         NaN         1292  ...        3          2   

                  Broker name Broker agent  \
0            RE/MAX Solutions          NaN   
1       RE/MAX Alliance Group          NaN   
2                DeLex Realty          NaN   
3              DPR Realty LLC          NaN   
4    Arizona Elite Properties          NaN   
..                        ...          ...   
434                       NaN          NaN   
435     RSP New Village Homes          NaN   
436                       NaN          NaN   
437     RSP New Village Homes          NaN   
438                       NaN          NaN   

                                   Street address      City    Zip  State  \
0                          93 N Cooper Rd UNIT 29  Chandler  85225     AZ   
1                         295 N Rural Rd UNIT 155  Chan

In [3]:
# Clean the dataframe by replacing NaN values with 0
df.fillna(0, inplace=True)

In [4]:
df

Property URL  MLS ID  \
0    https://www.zillow.com/homedetails/93-N-Cooper...     0.0   
1    https://www.zillow.com/homedetails/295-N-Rural...     0.0   
2    https://www.zillow.com/homedetails/1351-N-Plea...     0.0   
3    https://www.zillow.com/homedetails/1075-E-Chan...     0.0   
4    https://www.zillow.com/homedetails/295-N-Rural...     0.0   
..                                                 ...     ...   
434  https://www.zillow.com/community/village-at-ha...     0.0   
435  https://www.zillow.com/homedetails/960-E-Mulbe...     0.0   
436  https://www.zillow.com/community/willis-common...     0.0   
437  https://www.zillow.com/homedetails/972-E-Mulbe...     0.0   
438  https://www.zillow.com/community/village-at-ha...     0.0   

    Listing description Property type  Number of days on Zillow  \
0    Apartment for sale     Apartment                       0.0   
1    Apartment for sale     Apartment                       5.0   
2    Apartment for sale     Apartment                       4.0   
3    Apartment for sale     Apartment                      15.0   
4    Apartment for sale     Apartment                       0.0   
..                  ...           ...                       ...   
434    New construction       Unknown                      41.0   
435    New construction       Unknown                      74.0   
436    New construction       Unknown                       0.0   
437    New construction       Unknown                      74.0   
438    New construction       Unknown                      41.0   

     Property price (USD) Price was cut  Price cut amount (USD)  \
0                  289000            No                     0.0   
1                  389999            No                     0.0   
2                  398000            No                     0.0   
3                  225000            No                     0.0   
4                  299900           Yes                 35100.0   
..                    ...           ...                     ...   
434                607000            No                     0.0   
435                607000            No                     0.0   
436                580000            No                     0.0   
437                595000            No                     0.0   
438                595000            No                     0.0   

    Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                             0         1075  ...        2          2   
1                             0         1069  ...        2          2   
2                             0         1480  ...        3          2   
3                             0          705  ...        1          1   
4                    02/13/2024          979  ...        1          1   
..                          ...          ...  ...      ...        ...   
434                           0         1324  ...        3          2   
435                           0         1324  ...        3          2   
436                           0         2101  ...        3          3   
437                           0         1292  ...        3          2   
438                           0         1292  ...        3          2   

                  Broker name Broker agent  \
0            RE/MAX Solutions          0.0   
1       RE/MAX Alliance Group          0.0   
2                DeLex Realty          0.0   
3              DPR Realty LLC          0.0   
4    Arizona Elite Properties          0.0   
..                        ...          ...   
434                         0          0.0   
435     RSP New Village Homes          0.0   
436                         0          0.0   
437     RSP New Village Homes          0.0   
438                         0          0.0   

                                   Street address      City    Zip  State  \
0                          93 N Cooper Rd UNIT 29  Chandler  85225     AZ   
1                         295 N Rural Rd UNIT 155  Chan

In [5]:
# Import scikit-learn libraries including Random Forest Classifier
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier  # Changed from DecisionTreeClassifier to RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import plot_tree
import matplotlib.pyplot as plt

In [6]:
# Convert 'Price was cut' to a binary indicator (0 and 1)
df['Price was cut'] = df['Price was cut'].apply(lambda x: 1 if x == 'Yes' else 0)
df.tail()

Property URL  MLS ID  \
434  https://www.zillow.com/community/village-at-ha...     0.0   
435  https://www.zillow.com/homedetails/960-E-Mulbe...     0.0   
436  https://www.zillow.com/community/willis-common...     0.0   
437  https://www.zillow.com/homedetails/972-E-Mulbe...     0.0   
438  https://www.zillow.com/community/village-at-ha...     0.0   

    Listing description Property type  Number of days on Zillow  \
434    New construction       Unknown                      41.0   
435    New construction       Unknown                      74.0   
436    New construction       Unknown                       0.0   
437    New construction       Unknown                      74.0   
438    New construction       Unknown                      41.0   

     Property price (USD)  Price was cut  Price cut amount (USD)  \
434                607000              0                     0.0   
435                607000              0                     0.0   
436                580000              0                     0.0   
437                595000              0                     0.0   
438                595000              0                     0.0   

    Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
434                           0         1324  ...        3          2   
435                           0         1324  ...        3          2   
436                           0         2101  ...        3          3   
437                           0         1292  ...        3          2   
438                           0         1292  ...        3          2   

               Broker name Broker agent  \
434                      0          0.0   
435  RSP New Village Homes          0.0   
436                      0          0.0   
437  RSP New Village Homes          0.0   
438                      0          0.0   

                                   Street address      City    Zip  State  \
434  Plan 5 Plan, The Village at Hamilton Landing  Chandler  85286     AZ   
435                             960 E Mulberry Pl  Chandler  85286     AZ   
436                   Ancala Plan, Willis Commons  Chandler  85286     AZ   
437                             972 E Mulberry Pl  Chandler  85286     AZ   
438  Plan 3 Plan, The Village at Hamilton Landing  Chandler  85286     AZ   

    Country FederalInt  
434     USA       5.33  
435     USA       5.33  
436     USA       5.33  
437     USA       5.33  
438     USA       5.33  

[5 rows x 24 columns]

In [7]:
# Feature selection
# Selecting relevant features and the target variable
X = df[['Property price (USD)', 'Property type', 'Living area', 'Lot/land area', 'Bedrooms', 'Bathrooms', 'City', 'FederalInt']]
y = df['Price was cut']


In [8]:
# Handle the categorical variables
categorical_features = ['Property type', 'City']
label_encoders = {}

for col in categorical_features:
    le = LabelEncoder()
    X.loc[:,col] = le.fit_transform(X[col])
    label_encoders[col] = le

In [9]:
df

Property URL  MLS ID  \
0    https://www.zillow.com/homedetails/93-N-Cooper...     0.0   
1    https://www.zillow.com/homedetails/295-N-Rural...     0.0   
2    https://www.zillow.com/homedetails/1351-N-Plea...     0.0   
3    https://www.zillow.com/homedetails/1075-E-Chan...     0.0   
4    https://www.zillow.com/homedetails/295-N-Rural...     0.0   
..                                                 ...     ...   
434  https://www.zillow.com/community/village-at-ha...     0.0   
435  https://www.zillow.com/homedetails/960-E-Mulbe...     0.0   
436  https://www.zillow.com/community/willis-common...     0.0   
437  https://www.zillow.com/homedetails/972-E-Mulbe...     0.0   
438  https://www.zillow.com/community/village-at-ha...     0.0   

    Listing description Property type  Number of days on Zillow  \
0    Apartment for sale     Apartment                       0.0   
1    Apartment for sale     Apartment                       5.0   
2    Apartment for sale     Apartment                       4.0   
3    Apartment for sale     Apartment                      15.0   
4    Apartment for sale     Apartment                       0.0   
..                  ...           ...                       ...   
434    New construction       Unknown                      41.0   
435    New construction       Unknown                      74.0   
436    New construction       Unknown                       0.0   
437    New construction       Unknown                      74.0   
438    New construction       Unknown                      41.0   

     Property price (USD)  Price was cut  Price cut amount (USD)  \
0                  289000              0                     0.0   
1                  389999              0                     0.0   
2                  398000              0                     0.0   
3                  225000              0                     0.0   
4                  299900              1                 35100.0   
..                    ...            ...                     ...   
434                607000              0                     0.0   
435                607000              0                     0.0   
436                580000              0                     0.0   
437                595000              0                     0.0   
438                595000              0                     0.0   

    Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                             0         1075  ...        2          2   
1                             0         1069  ...        2          2   
2                             0         1480  ...        3          2   
3                             0          705  ...        1          1   
4                    02/13/2024          979  ...        1          1   
..                          ...          ...  ...      ...        ...   
434                           0         1324  ...        3          2   
435                           0         1324  ...        3          2   
436                           0         2101  ...        3          3   
437                           0         1292  ...        3          2   
438                           0         1292  ...        3          2   

                  Broker name Broker agent  \
0            RE/MAX Solutions          0.0   
1       RE/MAX Alliance Group          0.0   
2                DeLex Realty          0.0   
3              DPR Realty LLC          0.0   
4    Arizona Elite Properties          0.0   
..                        ...          ...   
434                         0          0.0   
435     RSP New Village Homes          0.0   
436                         0          0.0   
437     RSP New Village Homes          0.0   
438                         0          0.0   

                                   Street address      City    Zip  State  \
0                          93 N Cooper Rd UNIT 29  Chandler  85225     AZ   
1                         295 N Rural Rd UN

In [10]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [11]:
# Initializing and training the Decision Tree Classifier
clf = RandomForestClassifier(random_state=42)
clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [12]:
# Make predictions
y_pred = clf.predict(X_test)


In [13]:
# Evaluating the model
print("Accuracy Score:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))

Accuracy Score: 0.5454545454545454
Confusion Matrix:
 [[32 19]
 [21 16]]
Classification Report:
               precision    recall  f1-score   support

           0       0.60      0.63      0.62        51
           1       0.46      0.43      0.44        37

    accuracy                           0.55        88
   macro avg       0.53      0.53      0.53        88
weighted avg       0.54      0.55      0.54        88



# Initial Result of the Decision Tree Model

### Accuracy Score
The accuracy score of 0.545 (or 54.5%) indicates that the model correctly predicts the outcome (whether the price will be cut or not) in approximately 54.5% of the cases. This metric provides a general idea of the model's overall performance but doesn't detail its effectiveness in distinguishing between the two classes.



In [14]:
# Note: Random Forest trees might be too complex for practical visualization with plot_tree,
# especially for large numbers of trees. Consider summarizing tree characteristics or feature importances instead.
feature_importances = pd.Series(clf.feature_importances_, index=X.columns).sort_values(ascending=False)
print(feature_importances)

Living area             0.270309
Lot/land area           0.269627
Property price (USD)    0.268976
Bedrooms                0.067866
Property type           0.055870
Bathrooms               0.047477
City                    0.019875
FederalInt              0.000000
dtype: float64
